File: files/2025-07-06/CSEC 06072025.py
import pandas as pd

path = "files/2025-07-06/Challenge 38.xlsx"
input = pd.read_excel(path, usecols="B:D", skiprows=1, nrows=10)
test = pd.read_excel(path, usecols="F:H", skiprows=1, nrows=6).rename(columns=lambda col: col.replace('.1', ''))

result = (
    input.groupby(['Date', 'Store'], as_index=False)
    .agg(n=('Product', 'nunique'))
)
result['Product'] = result.groupby('Store')['n'].shift(1, fill_value=0)
result = result[['Date', 'Store', 'Product']]

print(result.equals(test))

File: files/2025-07-06/CSEC 06072025.R
library(tidyverse)
library(readxl)

path = "files/2025-07-06/Challenge 38.xlsx"
input = read_excel(path, range = "B2:D11")
test  = read_excel(path, range = "F2:H8")

result = input %>%
  summarise(n = n_distinct(Product),
            Product = paste0(Product, collapse = ", "),
            .by = c(Date,Store)) %>%
  mutate(lag = lag(n, default = 0), .by = Store) %>%
  select(Date, Store, Product = lag)

all.equal(result, test, check.attributes = FALSE)            
#> [1] TRUE